Deprecate `meta_keywords()`

exitflynn opened the pull request:
This was originally #6722 but I accidentally closed it since I didn't realise that force pushing would do that. 
Would close #6562.

exitflynn commented:
referring to Will's [comment](https://github.com/sunpy/sunpy/pull/6722#issuecomment-1404103904) on the original PR about the discussion in the dev call, I have tried to implement everything as mentioned:

> - Function should return a copy
> - Function should be deprecated
> - Dictionary should be all caps
> - Try making dictionary a MappingProxyType which is effectively an immutable dict: https://docs.python.org/3/library/types.html?highlight=mappingproxytype#types.MappingProxyType. There is a question of whether just making the meta keywords a dictionary effectively protects it or not.


I am unfamiliar with how to deprecate functions though, could really use some pointers/elaboration!

exitflynn commented:
> There is a question of whether just making the meta keywords a dictionary effectively protects it or not.

I checked and it doesn't. Also, neither does the current implementation. This is how I checked:

a = sunpy.map.header_helper.meta_keywords()
b = sunpy.map.header_helper.meta_keywords()
a['cunit1'] = 'R2-D2'
b

nabobalis commented:
> I am unfamiliar with how to deprecate functions though, could really use some pointers/elaboration!

https://docs.sunpy.org/en/latest/dev_guide/contents/public_api.html#deprecation-policy-and-breaking-changes

Should cover that hopefully.

exitflynn commented:
I looked around at some other PR's, is adding tests for this deprecation also required here?

exitflynn commented:
I looked around at some other PR's, is adding tests for this deprecation also required here?

nabobalis commented:
> I looked around at some other PR's, is adding tests for this deprecation also required here?

If the function is already tested, the deprecation will need to be caught to ensure it doesn't break the unit tests.

nabobalis commented:
The tests need updating due to the change in type.

nabobalis commented on sunpy/map/header_helper.py:
```diff
@@ -24,7 +28,7 @@ def meta_keywords():
          'cunit2': 'Units of the coordinate increments along naxis2 e.g. arcsec **required',
          ...
     """
-    return _map_meta_keywords
```
Can this not be a copy each time?

nabobalis commented on sunpy/map/header_helper.py:
```diff
@@ -24,7 +28,7 @@ def meta_keywords():
          ...
     """
-    return _map_meta_keywords
+    return MappingProxyType(_map_meta_keywords)
```
I think this counts as an API breaking change which we can't do. 
Can this not return a copy of _map_meta_keywords each time?

exitflynn commented:
I've changed the `meta_keywords()` function to return a copy of the dict instead of `MappingProxyType`. This also means that the tests won't break and therefore don't need to be updated, right? 
(forgive my commit once again traveling back in time, i've fixed the time synchronization issue on my machine for good now)

nabobalis commented on sunpy/map/header_helper.py:
```diff
@@ -7,10 +9,12 @@
 __all__ = ['meta_keywords', 'make_fitswcs_header', 'get_observer_meta', 'make_heliographic_header']
 
 
+@deprecated(since="4.1.2", message="We will be removing this", alternative="sunpy.map.header_helper.META_KEYWORDS")
```
```suggestion
@deprecated(since="5.0", message="Will be removed in 6.0", alternative="sunpy.map.header_helper.META_KEYWORDS")
```

nabobalis commented:
The tests will need to be changed since the function now raises a deprecation warning.

ayshih commented:
Before this PR can be accepted, I'd like there to be a consensus on whether this dictionary should continue to exist (see discussion in #6562.  It doesn't make sense to me to create a new module variable and have the deprecated function point to it, if we end up removing that module variable in short order.

nabobalis commented:
> Before this PR can be accepted, I'd like there to be a consensus on whether this dictionary should continue to exist (see discussion in #6562. It doesn't make sense to me to create a new module variable and have the deprecated function point to it, if we end up removing that module variable in short order.

Fair point. 

How about we remove module variable for now and add it back if we want to expose it?

hayesla commented:
I think this can be deprecated and then removed in place of proper documentation on the docs (i.e. description of meta keywords to FITS-WCS), which I'm aware isn't specific to sunpy, but certainly useful to the community

nabobalis commented:
We discussed this in the community meeting today:

1. Fix the function to return a copy
2. Deprecate it
3. Remove the parts of the map.rst file that mentions this function and the documention test.

exitflynn commented:
does ^including the meta_keywords in the docs like ^that work or should i go with something like [sphinx tables](https://sublime-and-sphinx-guide.readthedocs.io/en/latest/tables.html) instead?

nabobalis commented on sunpy/map/tests/test_header_helper.py:
```diff
@@ -61,11 +61,6 @@ def hpc_coord_notime():
     return SkyCoord(0*u.arcsec, 0*u.arcsec, frame=frames.Helioprojective)
 
 
-def test_metakeywords():
-    meta = sunpy.map.meta_keywords()
-    assert isinstance(meta, dict)
```
This test should stay.

nabobalis commented on sunpy/map/header_helper.py:
```diff
@@ -382,6 +386,8 @@ def _set_instrument_meta(meta_wcs, instrument, telescope, observatory, detector,
     'Matrix element CDi_j describing the rotation required to align solar North with the top of the image.'
 }
 
+META_KEYWORDS = MappingProxyType(_map_meta_keywords)
```
This should go.

nabobalis commented on docs/intro/maps.rst:
```diff
@@ -491,20 +491,51 @@ The meta information informs `sunpy.map.Map` of the correct coordinate informati
      ...
 
-The utility function :func:`~sunpy.map.header_helper.make_fitswcs_header` will return a header with the appropriate FITS keywords once the Map data array and an `astropy.coordinates.SkyCoord` or `sunpy.coordinates.frames` is provided.
+**sunpy** provides a Map header helper function to assist in creating a header that contains the correct meta information.
+This includes a utility function :func:`~sunpy.map.header_helper.make_fitswcs_header` that will return a header with the appropriate FITS keywords once the Map data array and an `astropy.coordinates.SkyCoord` or `sunpy.coordinates.frames` is provided.
```
```suggestion
``sunpy`` provides a helper, :func:`~sunpy.map.header_helper.make_fitswcs_header`, to assist in creating a header that contains the correct metadata.
This will return a header with the appropriate FITS keywords once the Map data array and an `astropy.coordinates.SkyCoord` or `sunpy.coordinates.frames` is provided.
```

nabobalis commented on docs/intro/maps.rst:
```diff
@@ -491,20 +491,51 @@ The meta information informs `sunpy.map.Map` of the correct coordinate informati
+`CD1_2`: Matrix element CDi_j describing the rotation required to align solar North with the top of the image.
+`CD2_1`: Matrix element CDi_j describing the rotation required to align solar North with the top of the image.
+`CD2_2`: Matrix element CDi_j describing the rotation required to align solar North with the top of the image.
      ...
```
```suggestion
```

nabobalis commented:
> does including the meta_keywords in the docs like ^that work or should i go with something like [sphinx tables](https://sublime-and-sphinx-guide.readthedocs.io/en/latest/tables.html) instead?

I think a formatted table would be nice.

nabobalis commented:
You will need to update the unit test to catch the deprecation warning to prevent the test from failing.

nabobalis commented on docs/intro/maps.rst:
```diff
@@ -491,20 +491,50 @@ The meta information informs `sunpy.map.Map` of the correct coordinate informati
-     ...
-
-The utility function :func:`~sunpy.map.header_helper.make_fitswcs_header` will return a header with the appropriate FITS keywords once the Map data array and an `astropy.coordinates.SkyCoord` or `sunpy.coordinates.frames` is provided.
+All the meta keywords used when creating a Map along with their description are listed below.
```
```suggestion
All the metadata keywords that a Map will parse along with their description are listed below:
```

nabobalis commented on docs/intro/maps.rst:
```diff
@@ -491,20 +491,50 @@ The meta information informs `sunpy.map.Map` of the correct coordinate informati
-The utility function :func:`~sunpy.map.header_helper.make_fitswcs_header` will return a header with the appropriate FITS keywords once the Map data array and an `astropy.coordinates.SkyCoord` or `sunpy.coordinates.frames` is provided.
+All the meta keywords used when creating a Map along with their description are listed below.
+
+`cunit1`: Units of the coordinate increments along naxis1 e.g. arcsec \**required
```
You will need to remove the ` on either side of the variable name and remove the `\**`

nabobalis commented on sunpy/map/header_helper.py:
```diff
@@ -1,3 +1,4 @@
+
```
```suggestion
```

nabobalis commented on sunpy/map/header_helper.py:
```diff
@@ -24,7 +27,7 @@ def meta_keywords():
          'cunit2': 'Units of the coordinate increments along naxis2 e.g. arcsec **required',
```
I think this example can go as well. 

Or we need to catch the waring but I forgot how to do that for doctests.

exitflynn replied:
i think a `# doctest: +SKIP` should do that?

exitflynn replied:
regardless, i guess we should just remove the example since we're deprecating the function

nabobalis commented on sunpy/map/header_helper.py:
```diff
@@ -7,10 +8,12 @@
 __all__ = ['meta_keywords', 'make_fitswcs_header', 'get_observer_meta', 'make_heliographic_header']
 
 
+@deprecated(since="5.0", message="Will be removed in 6.0", alternative="sunpy.map.header_helper.META_KEYWORDS")
```
```suggestion
@deprecated(since="5.0", message="Unused and will be removed in 6.0")
```

nabobalis commented on docs/intro/maps.rst:
```diff
@@ -491,20 +491,50 @@ The meta information informs `sunpy.map.Map` of the correct coordinate informati
-The utility function :func:`~sunpy.map.header_helper.make_fitswcs_header` will return a header with the appropriate FITS keywords once the Map data array and an `astropy.coordinates.SkyCoord` or `sunpy.coordinates.frames` is provided.
+All the metadata keywords that a Map will parse along with their description are listed below:
+
+cunit1: Units of the coordinate increments along naxis1 e.g. arcsec (required)
```
This currently looks like: https://sunpy--6743.org.readthedocs.build/en/6743/intro/maps.html#creating-custom-maps

Hopefully turning it into a table will look nicer.

exitflynn replied:
i see, totally, will get on that :+1:

nabobalis replied:
I think this should probably go at the end of the page as well.

nabobalis commented on docs/intro/maps.rst:
```diff
@@ -491,20 +491,72 @@ The meta information informs `sunpy.map.Map` of the correct coordinate informati
-The utility function :func:`~sunpy.map.header_helper.make_fitswcs_header` will return a header with the appropriate FITS keywords once the Map data array and an `astropy.coordinates.SkyCoord` or `sunpy.coordinates.frames` is provided.
+All the metadata keywords that a Map will parse along with their description are listed below:
+
+.. list-table:: Frozen Delights!
```
Did you mean to commit this?

exitflynn replied:
this happened by mistake, will fix this

exitflynn commented:
that ought to do it

nabobalis commented:
The unit test needs the SunpyDeprecationWarning to be checked:
```
______________________________ test_metakeywords _______________________________
[gw0] linux -- Python 3.11.1 /home/runner/work/sunpy/sunpy/.tox/py311/bin/python

    def test_metakeywords():
>       meta = sunpy.map.meta_keywords()

../../.tox/py311/lib/python3.11/site-packages/sunpy/map/tests/test_header_helper.py:65: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

args = (), kwargs = {}
category = <class 'sunpy.util.exceptions.SunpyDeprecationWarning'>

    def deprecated_func(*args, **kwargs):
        if pending:
            category = pending_warning_type
        else:
            category = warning_type
    
>       warnings.warn(message, category, stacklevel=2)
E       sunpy.util.exceptions.SunpyDeprecationWarning: Unused and will be removed in 6.0
```

nabobalis commented on docs/intro/maps.rst:
```diff
@@ -505,6 +505,11 @@ This includes a :func:`~sunpy.map.header_helper.meta_keywords` function that wil
 The utility function :func:`~sunpy.map.header_helper.make_fitswcs_header` will return a header with the appropriate FITS keywords once the Map data array and an `astropy.coordinates.SkyCoord` or `sunpy.coordinates.frames` is provided.
+All the metadata keywords that a Map will parse along with their description are listed in the :ref:`Meta Keywords Table` at the end of this page.
+
+
```
```suggestion
```

nabobalis commented on docs/intro/maps.rst:
```diff
@@ -649,3 +654,91 @@ From these header MetaDict's that are generated, we can now create a custom map:
 
     >>> my_map = sunpy.map.Map(data, header)
+
+
```
```suggestion
```

nabobalis commented on docs/intro/maps.rst:
```diff
@@ -505,6 +505,11 @@ This includes a :func:`~sunpy.map.header_helper.meta_keywords` function that wil
+All the metadata keywords that a Map will parse along with their description are listed in the :ref:`Meta Keywords Table` at the end of this page.
+
+
+``sunpy`` provides a helper, :func:`~sunpy.map.header_helper.make_fitswcs_header`, to assist in creating a header that contains the correct metadata.
```
```suggestion
``sunpy`` provides :func:`~sunpy.map.header_helper.make_fitswcs_header`, to assist in creating a FITS header that contains the correct metadata.
```

ayshih commented on changelog/6743.deprecation.rst:
```diff
@@ -0,0 +1,2 @@
+Using `~sunpy.map.header_helper.meta_keywords()` is deprecated.
+Use `~sunpy.map.header_helper.META_KEYWORDS` as a `MappingProxyType` dictionary of the meta keywords.
```
```suggestion
```

nabobalis commented on sunpy/map/header_helper.py:
```diff
@@ -7,10 +7,12 @@
 __all__ = ['meta_keywords', 'make_fitswcs_header', 'get_observer_meta', 'make_heliographic_header']
 
 
+@deprecated(since="5.0", message="Unused and will be removed in 6.0")
```
The unit test for this function needs to use `pytest.warns` to catch and check for the deprecation warning.

nabobalis commented on changelog/6743.deprecation.rst:
```diff
@@ -0,0 +1 @@
+Using `~sunpy.map.header_helper.meta_keywords()` is deprecated.
```
```suggestion
Using `~sunpy.map.header_helper.meta_keywords` is deprecated.
Please see :ref:`Meta Keywords Table` for the list of metadata keywords used by `~sunpy.map.Map`.
```

exitflynn replied:
aahh thanks a lot there, i was having an unbelievably hard time figuring out how to do this (and the tests) :sob: 
was overthinking it, this looks very obvious in hindsight!

nabobalis replied:
We will need to see if this works first.

exitflynn replied:
it works! 🥳

exitflynn commented:
i think this should be good to go now :thinking:

nabobalis commented:
> i think this should be good to go now 🤔

Once someone reviews it.

nabobalis commented:
Thanks again @exitflynn

exitflynn commented:
:sparkles: :partying_face: 
a really big thanks to everyone for reviewing!
